rfcs/0000-safe-transmute.md

```diff
@@ -0,0 +1,2160 @@
+    }
+
+    // introduce a private module to avoid `private_in_public` error (E0446):
+    pub(crate) mod private {
+        #[repr(C)]
+        pub struct NonEmptySliceInner<T> {
+            pub data: *const T,
+            pub len: usize,
+        }
+    }
+
+}
+```
+With this simplified definition of constructability, it is possible for a third-party to define a *safe* constructor of `NonEmptySlice` that produces a value which is *unsafe* to use:
+```rust
+pub evil_constructor<T>(src: T) -> NonEmptySlice<u8>
+where
+    T: TransmuteInto<NonEmptySlice<u8>, NeglectStability>,
+{
+    src.transmute_into()
+}
+
+evil_constructor(0u128).first() // muahaha!
+```
+
+The above code is "safe" because our simplified definition of constructability fails to recognize this pattern of encapsulation, and because `NeglectStability` is a `SafeTransmutationOption`.
+
+The intent of `NeglectStability` is to permit the safe transmutation of types that predate the stabilization of the stability declaration traits. It also provides a convenient escape-hatch for type authors to neglect the stability of transmutations of their *own* types, without sacrificing safety. `NeglectStability` is a `SafeTransmutationOption` because, in principle, neglecting stability does not diminish safety. Our simplified definition of constructability violates this principle.
+
+By temporarily sacrificing these goals, we may preserve safety solely...
+
+##### ...at the Cost of `NeglectStability`
+We may preserve safety by demoting `NeglectStability` to `UnsafeTransmutationOption`-status.
+
+In doing so, a third-party is forced to resort to an `unsafe` transmutation to construct `NonEmptySlice`; e.g.:
+
+```rust
+pub evil_constructor<T>(src: T) -> NonEmptySlice<u8>
+where
+    T: TransmuteInto<NonEmptySlice<u8>, NeglectStability>,
+{
+    // unsafe because we `NeglectStability`
+    unsafe { src.unsafe_transmute_into() }
+}
+```
+
+Demoting `NeglectStability` to unsafe-status does not stop type authors from opting-in to stable (and thus safe) transmutations; e.g., with `derive(PromiseTransmutableFrom)`:
+```rust
+pub mod crate_a {
+
+    #[derive(PromiseTransmutableFrom)]
+    #[repr(C)]
+    pub struct NonEmptySlice<T>(pub private::NonEmptySliceInner<T>);
+
+    impl<T> NonEmptySlice<T> {
+        pub fn from_array<const N: usize>(arr: &[T; N], len: usize) -> Self {
+            assert!(len <= N && len > 0);
+            Self(
+                private::NonEmptySliceInner {
+                    data: arr as *const T,
+                    len,
+                }
+            )
+        }
+
+        pub fn first(&self) -> &T {
+            unsafe { &*self.0.data }
+        }
+    }
+
+    // introduce a private module to avoid `private_in_public` error (E0446):
+    pub(crate) mod private {
+        #[derive(PromiseTransmutableFrom)]
+        #[repr(C)]
+        pub struct NonEmptySliceInner<T> {
+            pub data: *const T,
+            pub len: usize,
+        }
+    }
+
+}
```

Comment:
```suggestion
pub mod crate_a {

    #[derive(PromiseTransmutableFrom)]
    #[repr(C)]
    pub struct NonEmptySlice<'a, T>(pub private::NonEmptySliceInner<'a, T>);

    impl<'a, T> NonEmptySlice<'a, T> {
        pub fn from_array<const N: usize>(arr: &'a [T; N], len: usize) -> Self {
            assert!(len <= N && len > 0);
            Self(
                private::NonEmptySliceInner {
                    data: arr as *const T,
                    len,
                    lifetime: core::marker::PhantomData,
                }
            )
        }

        pub fn first(&self) -> &'a T {
            unsafe { &*self.0.data }
        }
    }

    // introduce a private module to avoid `private_in_public` error (E0446):
    pub(crate) mod private {
        #[derive(PromiseTransmutableFrom)]
        #[repr(C)]
        pub struct NonEmptySliceInner<'a, T> {
            pub data: *const T,
            pub len: usize,
            pub lifetime: core::marker::PhantomData<&'a ()>,
        }
    }

}
```